pkg/gds/admin.go

```diff
@@ -149,22 +160,193 @@ func (s *Admin) setupRoutes() (err error) {
 	return nil
 }
 
+// Set the maximum age of authentication protection cookies.
+const protectAuthenticateMaxAge = time.Minute * 10
+
+// ProtectAuthenticate prepares the front-end for submitting a login token by setting
+// the double cookie tokens for CSRF protection. The front-end should call this before
+// posting credentials from Google.
+func (s *Admin) ProtectAuthenticate(c *gin.Context) {
+	expiresAt := time.Now().Add(protectAuthenticateMaxAge).Unix()
+	if err := admin.SetDoubleCookieTokens(c, expiresAt); err != nil {
+		c.JSON(http.StatusInternalServerError, admin.ErrorResponse("could not set cookies"))
+		return
+	}
+	c.JSON(http.StatusOK, &admin.Reply{Success: true})
+}
+
+// Authenticate expects a Google OAuth JWT token that is verified by the server. Once
+// verified, the JWT claims are authenticated against the server. Provided valid claims,
+// the server will issue access and referesh tokens that the client should submit in the
+// Authorization header for all future requests. This method also resets the CSRF double
+// cookies to ensure that max-age matches the duration of the refresh tokens.
 func (s *Admin) Authenticate(c *gin.Context) {
-	if err := admin.SetDoubleCookieTokens(c, time.Now().Add(time.Minute*10).Unix()); err != nil {
+	var (
+		err          error
+		in           *admin.AuthRequest
+		out          *admin.AuthReply
+		claims       *idtoken.Payload
+		accessToken  *jwt.Token
+		refreshToken *jwt.Token
+	)
+
+	// Parse incoming JSON data from the client request
+	in = new(admin.AuthRequest)
+	if err = c.ShouldBind(&in); err != nil {
+		log.Warn().Err(err).Msg("could not bind request")
+		c.JSON(http.StatusBadRequest, admin.ErrorResponse(err))
+		return
+	}
+
+	// Check that a credential was posted
+	if in.Credential == "" {
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("invalid credentials"))
+		return
+	}
+
+	// Validate the credential with Google
+	if claims, err = idtoken.Validate(c.Request.Context(), in.Credential, s.conf.Audience); err != nil {
+		log.Warn().Err(err).Msg("invalid credentials used for authentication")
+		c.JSON(http.StatusUnauthorized, admin.ErrorResponse("invalid credentials"))
+		return
+	}
```

Comment:
This function ensures that the token is valid, actually from Google, not expired, and that it matches our Google Cloud project credentials.

Reply:
Still TODO: validate our own claims; e.g. is this a user with a valid admin domain?

Reply:
The Google JWT token contains an `aud` value (JWT for "Audience") that is from the API credentials set up step in the Google Project. Both the front-end and the backend need to know this variable in order to engage the OAuth process. Audience here just verifies that the front end and the backend are using the same credentials, it does not provide any extra security.